### DISCUSSION TITLE
Xunit generates a global using. How to stop it?

### DISCUSSION THREAD

**Discussion by CaseyHofland** [xunit/xunit#3039] [Question] [answered]
Xunit generates a global using for me.

```csharp
// <auto-generated/>
global using global::Xunit;
```

Really helpful so I don't have to put `using Xunit;` in all my test files! And that's great and all, but I am testing on multiple frameworks, in my case `<TargetFrameworks>net8.0;netcoreapp3.1</TargetFrameworks>`. The second one doesn't support global usings, but the first one generates them all the same. I just want to disable it altogether.

I couldn't find anything about this in the [configuration file](https://xunit.net/docs/configuration-files).

Naturally, my csproj has its own global usings disabled (`<ImplicitUsings>disable</ImplicitUsings>`) but the package reference doesn't respect that setting it seems.

**Comment by CaseyHofland** [comment]
Alright, after some testing I figured it out. When I was creating the test framework, it was asking me what framework to use and I selected .Net 8 as the default. This made it install `xunit 2.5.3`. If I selected .Net Core 3.1 it would install `xunit 2.4.0` by default. This version doesn't generate global usings.

Guess it's good practice to always install from the lowest version you support upwards. My bad, this is my first project where I target multiple frameworks.

I'm leaving this here in case anybody runs into the same issue.

![image](https://github.com/user-attachments/assets/d7131106-87e0-4ae1-8e12-b5bf4dd9c02b)

**Reply by bradwilson** [reply, marked as answer]
FWIW, we don't control that template, Microsoft does. If there are issues/complaints with it, they should be reported to Microsoft.

I believe in general (and not specific to test projects), they swap in/out features like global usings based on the chosen target framework, as you've discovered. Global usings are completely orthogonal to what version of xUnit.net you use. They are, in fact, a compiler feature and not a runtime feature, so they're doing it based on what they consider to by the minimum supported _compiler_ version for the chosen framework. They've probably done this by shipping older templates (hence, the older xUnit.net version in it) rather than making the templates dynamically enable features like global usings.

**Reply by bradwilson** [reply]
> Naturally, my csproj has its own global usings disabled (`<ImplicitUsings>disable</ImplicitUsings>`) but the package reference doesn't respect that setting it seems.

That's correct, this setting is per-project.

> in my case `<TargetFrameworks>net8.0;netcoreapp3.1</TargetFrameworks>`. The second one doesn't support global usings, but the first one generates them all the same.

As mentioned above, this is a compiler option, not a runtime option; it's just that each target framework comes with its own 'default' compiler version. You _could_ absolutely use global usings with both just by setting a higher C# compiler version via [`<LangVersion>`](https://learn.microsoft.com/dotnet/csharp/language-reference/configure-language-version) (even though, in this case, you'd prefer for it to be globally off; I'm just letting any future readers know that this is possible).
